File: src/A15.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

import static java.lang.Math.abs;

public class A15 {
    public static void main(String...a) throws IOException {
        final List<Scanner> scanners = Files.lines(Paths.get("s-")).map(Scanner::new).toList();
        final boolean[] b = new boolean[85];
        scanners.forEach(s -> s.fill(b, 10, 5));
        int s = 0;
        for (boolean value : b) {
            if (value) s++;
        }
        System.out.println(s);
    }
}

class Scanner {
    final int sx, sy, bx, by, d;
    public Scanner(String s) {
        final int q = s.indexOf('=') + 1;
        final int w = s.indexOf(',', q);
        sx = Integer.parseInt(s.substring(q, w));
        final int e = s.indexOf('=', w) + 1;
        final int r = s.indexOf(':', e);
        sy = Integer.parseInt(s.substring(e, r));
        final int t = s.indexOf('=', r) + 1;
        final int y = s.indexOf(',', t);
        bx = Integer.parseInt(s.substring(t, y));
        final int u = s.indexOf('=', y) + 1;
        by = Integer.parseInt(s.substring(u));
        d = abs(sx - bx) + abs(sy - by);
    }

    public void fill(boolean[] b, int row, int offset) {
        System.out.println(this);
        final int t = d - abs(sy - row);
        if (t < 0) return;
        for (int i = offset + sx - t; i <= offset + sx + t; i++) {
            b[i] = true;
        }
        for (boolean value : b) {
            System.out.print(value ? '#' : '.');
        }
        System.out.println();
    }

    @Override
    public String toString() {
        return "[" + sx + "," + sy + "]:[" + bx + "," + by + "]=" + d;
    }
}

File: src/A15.kt
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*
import kotlin.math.abs

object A15 {
    @Throws(IOException::class)
    @JvmStatic
    fun main(a: Array<String>) {
        val scanners = Files.lines(Paths.get("s")).map { s: String -> Scanner(s) }.toList()
        val intervals = scanners.stream().map { s -> s.createInterval(2000000) }
            .filter { obj -> Objects.nonNull(obj) }.sorted(Comparator.comparingInt { i -> i!!.s }).toList()

        println(mergeIntervals(intervals).stream().mapToInt { obj -> obj!!.size }.sum() - 1)

        for (y in 0..4000000) {
            val intervalList = mergeIntervals(scanners.stream().map { s -> s.createInterval(y) }
                .filter { obj -> Objects.nonNull(obj) }.sorted(Comparator.comparingInt { i -> i!!.s }).toList())
            if (intervalList.size != 1) {
                println((intervalList[0].e + 1L) * 4000000 + y)
            }
        }
    }

    private fun mergeIntervals(arr: List<Interval?>): List<Interval> {
        val stack = Stack<Interval>()
        stack.push(arr[0])

        for (i in 1 until arr.size) {
            val current = arr[i]
            val top = stack.peek()

            if (top!!.e < current!!.s)
                stack.push(current)
            else if (top.e < current.e) {
                top.e = current.e
                stack.pop()
                stack.push(top)
            }
        }
        return stack
    }
}

class Interval(var s: Int, var e: Int) {
    val size: Int
        get() = 1 + e - s
}

internal class Scanner(s: String) {
    private val sx: Int
    private val sy: Int
    private val r: Int

    init {
        val q = s.indexOf('=') + 1
        val w = s.indexOf(',', q)
        sx = s.substring(q, w).toInt()
        val e = s.indexOf('=', w) + 1
        val h = s.indexOf(':', e)
        sy = s.substring(e, h).toInt()
        val t = s.indexOf('=', h) + 1
        val y = s.indexOf(',', t)
        val bx = s.substring(t, y).toInt()
        val u = s.indexOf('=', y) + 1
        val by = s.substring(u).toInt()
        r = abs(sx - bx) + abs(sy - by)
    }

    fun createInterval(row: Int): Interval? {
        val t = r - abs(sy - row)
        if (t < 0) return null
        return Interval(sx - t, sx + t)
    }
}
